provisionersdk/session.go

```diff
@@ -0,0 +1,307 @@
+package provisionersdk
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"cdr.dev/slog"
+
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/provisionersdk/proto"
+)
+
+// ReadmeFile is the location we look for to extract documentation from template
+// versions.
+const ReadmeFile = "README.md"
+
+// protoServer is a wrapper that translates the dRPC protocol into a Session with method calls into the Server.
+type protoServer struct {
+	server Server
+	opts   ServeOptions
+}
+
+func (p *protoServer) Session(stream proto.DRPCProvisioner_SessionStream) error {
+	s := &Session{
+		Logger: p.opts.Logger,
+		stream: stream,
+		server: p.server,
+	}
+	sessDir := fmt.Sprintf("Session%p", s)
```

Comment:
We need something unique to the Session, for the lifetime of the Session.  The pointer is a simple and easy choice.

Each provisioner needs to be run with a unique work directory so that they don't trample on each others caches, so there is no need to qualify the pointer with a PID.

The PID might make things easier to find... if you know the PID, which is rarely the case _after_ you have a crash.